packages/controllers/src/permissions/PermissionController.test.ts

```diff
@@ -0,0 +1,4014 @@
+                },
+              ] as any,
+            },
+          },
+        }),
+      ).toThrow(
+        new errors.ForbiddenCaveatError(
+          CaveatTypes.filterArrayResponse,
+          origin,
+          PermissionNames.wallet_doubleNumber,
+        ),
+      );
+    });
+
+    it('throws if the permission validator throws', () => {
+      const controller = getDefaultPermissionController();
+      const origin = 'metamask.io';
+
+      expect(() =>
+        controller.grantPermissions({
+          subject: { origin },
+          approvedPermissions: {
+            [PermissionNames.wallet_noop2]: {
+              caveats: [
+                { type: CaveatTypes.filterArrayResponse, value: ['foo'] },
+              ],
+            },
+          },
+        }),
+      ).toThrow(new Error('noop permission validation failed'));
+    });
+  });
+
+  describe('requestPermissions', () => {
+    it('requests a permission', async () => {
+      const options = getPermissionControllerOptions();
+      const { messenger } = options;
+      const origin = 'metamask.io';
+
+      const callActionSpy = jest
+        .spyOn(messenger, 'call')
+        .mockImplementationOnce((async (...args: any) => {
+          const [, { requestData }] = args;
+          return {
+            metadata: { ...requestData.metadata },
+            permissions: { ...requestData.permissions },
+          };
+        }) as any);
+
+      const controller = getDefaultPermissionController(options);
+      expect(
+        await controller.requestPermissions(
+          { origin },
+          {
+            [PermissionNames.wallet_getSecretArray]: {},
+          },
+        ),
+      ).toMatchObject([
+        {
+          [PermissionNames.wallet_getSecretArray]: getPermissionMatcher(
+            PermissionNames.wallet_getSecretArray,
+            null,
+            origin,
+          ),
+        },
+        { id: expect.any(String), origin },
+      ]);
+
+      expect(callActionSpy).toHaveBeenCalledTimes(1);
+      expect(callActionSpy).toHaveBeenCalledWith(
+        'ApprovalController:addRequest',
+        {
+          id: expect.any(String),
+          origin,
+          requestData: {
+            metadata: { id: expect.any(String), origin },
+            permissions: { [PermissionNames.wallet_getSecretArray]: {} },
+          },
+          type: MethodNames.requestPermissions,
+        },
+        true,
+      );
+    });
+
+    it('requests multiple permissions', async () => {
+      const options = getPermissionControllerOptions();
+      const { messenger } = options;
+      const origin = 'metamask.io';
+
+      const callActionSpy = jest
+        .spyOn(messenger, 'call')
+        .mockImplementationOnce((async (...args: any) => {
+          const [, { requestData }] = args;
+          return {
+            metadata: { ...requestData.metadata },
+            permissions: { ...requestData.permissions },
+          };
+        }) as any);
```

Comment:
The cast is not needed here either:
```suggestion
        .mockImplementationOnce(async (...args: any) => {
          const [, { requestData }] = args;
          return {
            metadata: { ...requestData.metadata },
            permissions: { ...requestData.permissions },
          };
        });
```